doc/source/serve/deployment-graph.md

```diff
@@ -0,0 +1,401 @@
+
++++
+
+## Full End to End Example
+
+Let's put the concepts together to a full runnable DAG example including the following attributes where each node is empowered by a __[serve deployment](https://docs.ray.io/en/master/serve/core-apis.html#core-api-deployments)__:
+
+- All nodes in the deployment graph naturally forms a DAG structure.
+- A node could use or call into other nodes in the deployment graph.
+- Deployment graph has mix of class and function as nodes.
+- Same input or output can be used in multiple nodes in the DAG.
+- A node might access partial user input.
+- Control flow is used where dynamic dispatch happens with respect to input value.
+- Same class can be constructed, or function bound with different args that generates multiple distinct nodes in DAG.
+- A node can be called either sync or async.
+
++++
+
+![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)
+
+```{code-cell} ipython3
+import ray
+from ray import serve
+from ray.serve.pipeline.generate import DeploymentNameGenerator
+
+if ray.is_initialized():
+    serve.shutdown()
+    DeploymentNameGenerator.reset()
+    ray.shutdown()
+
+ray.init(num_cpus=16)
+serve.start()
+
+### Setup clean ray cluster with new serve instance ###
+
+import time
+import asyncio
+import requests
+import starlette
+
+from ray.experimental.dag.input_node import InputNode
+
+@serve.deployment
+async def preprocessor(input_data: str):
+    """Simple feature processing that converts str to int"""
+    time.sleep(0.1) # Manual delay for blocking computation
+    return int(input_data)
+
+@serve.deployment
+async def avg_preprocessor(input_data):
+    """Simple feature processing that returns average of input list as float."""
+    time.sleep(0.15) # Manual delay for blocking computation
+    return sum(input_data) / len(input_data)
+
+@serve.deployment
+class Model:
+    def __init__(self, weight: int):
+        self.weight = weight
+
+    async def forward(self, input: int):
+        time.sleep(0.3) # Manual delay for blocking computation 
+        return f"({self.weight} * {input})"
+
+
+@serve.deployment
+class Combiner:
+    def __init__(self, m1: Model, m2: Model):
+        self.m1 = m1
+        self.m2 = m2
+
+    async def run(self, req_part_1, req_part_2, operation):
+        # Merge model input from two preprocessors  
+        req = f"({req_part_1} + {req_part_2})"
+        
+        # Submit to both m1 and m2 with same req data in parallel
+        r1_ref = self.m1.forward.remote(req)
+        r2_ref = self.m2.forward.remote(req)
+        
+        # Async gathering of model forward results for same request data
+        rst = await asyncio.gather(*[r1_ref, r2_ref])
+        
+        # Control flow that determines runtime behavior based on user input
+        if operation == "sum":
+            return f"sum({rst})"
+        else:
+            return f"max({rst})"
+        
+@serve.deployment(num_replicas=2)
+class DAGDriver:
```

Comment:
let's just use the built-in driver now?

Reply:
i tried but i didn't work due to the input_schema thing :/  it also some some rough edges of user input because ray dag's InputNode accepts .remote(1, 2, x=3) but it didn't match default args of DAGDriver.

Reply:
i had it earlier while putting up the example: https://github.com/ray-project/ray/issues/23484